### PR TITLE
Create a release promotion action

### PR DESCRIPTION
**Pull request by ahal**
### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: taskcluster/app_services_taskgraph/transforms/deps_complete.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 from taskgraph.transforms.base import TransformSequence
 
 transforms = TransformSequence()
+alerts = TransformSequence()
```
[the remark this answers is not in the document]
Comment (ahal) [reply taskcluster/app_services_taskgraph/transforms/deps_complete.py:8]: It's used e.g, here: https://github.com/mozilla/application-services/pull/5501/files#diff-3fd946aa80e283c20e3fc25009310bc8360f75aa76a03c9a34bbce67e67d7c18R9

[the remark this answers is not in the document]
Comment (ahal) [reply taskcluster/app_services_taskgraph/transforms/deps_complete.py:8]: I guess I could have pulled that out into a new transforms file, but I was feeling a bit lazy. Not opposed to doing so if folks want

**Review by bendk — DISMISSED** [review]
I don't have enough knowledge to comment on the `release_promotion.py` script, but the other changes look good to me.

I look forward to iterating on this, I think this could improve some of the code I recently added for nightlies.

File: taskcluster/app_services_taskgraph/actions/release_promotion.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    context=[],
+    available=lambda p: p["head_repository"] in RELEASE_PROMOTION_PROJECTS,
+    schema=lambda graph_config: {
+        "type": "object",
+        "properties": {
+            "release_promotion_flavor": {
+                "type": "string",
+                "description": "The flavor of release promotion to perform.",
+                "default": "promote",
+                "enum": sorted(graph_config["release-promotion"]["flavors"].keys()),
```
Comment (jcristau) [inline taskcluster/app_services_taskgraph/actions/release_promotion.py:37]: nit: no need for `.keys()`?

**Review by jcristau — COMMENTED** [review]
One thing that confuses me while reviewing this is that `taskgraph target-graph  -p taskcluster/test/params/release-branch-push.yml` includes signing and beetmover tasks, apparently related to the `branch-build` stuff which I don't understand, but seems wrong.

**Comment by bendk** [comment]
> One thing that confuses me while reviewing this is that `taskgraph target-graph -p taskcluster/test/params/release-branch-push.yml` includes signing and beetmover tasks, apparently related to the `branch-build` stuff which I don't understand, but seems wrong.

That seems wrong to me as well, and I don't think `branch-build` should come into play here.  I think this should either be fixed in this PR or I can fix it in a follow-up.

File: taskcluster/app_services_taskgraph/target_tasks.py [hunk 3]
```diff
@@ -43,3 +45,41 @@ def filter(task):
...
+@_target_task("ship")
+def target_tasks_ship(full_task_graph, parameters, graph_config):
+    filtered_for_candidates = target_tasks_promote(
+        full_task_graph,
+        parameters,
+        graph_config,
+    )
+    return filter_release_promotion(
+        full_task_graph, filtered_for_candidates, shipping_phase="ship"
+    )
```
Comment (bendk) [inline taskcluster/app_services_taskgraph/target_tasks.py:85]: Does this mean that we'll run the `promote` tasks again in the `ship` phase or will they be cached somehow?

**Comment by ahal** [comment]
> One thing that confuses me while reviewing this is that `taskgraph target-graph -p taskcluster/test/params/release-branch-push.yml` includes signing and beetmover tasks, apparently related to the `branch-build` stuff which I don't understand, but seems wrong.

Good catch, I'll investigate.. This is unrelated to the current PR though as I can reproduce on main as well. That reminds me that I should add some relpro params files too.

**Pull request merged**